### PR TITLE
[rtsan] Add ioctl interceptor

### PR DESCRIPTION
**Pull request by cjappl**

### CODE REVIEW

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors_posix.cpp [hunk 1]
```diff
@@ -373,6 +375,57 @@ class RtsanOpenedFileTest : public RtsanFileTest {
...
+}
+
+TEST_F(RtsanOpenedFileTest, IoctlBehavesWithOutputArg) {
+  int arg{};
+  ioctl(GetOpenFd(), FIONREAD, &arg);
+
+  EXPECT_THAT(arg, Ge(0));
+}
+
+TEST(TestRtsanInterceptors, IoctlBehavesWithOutputPointer) {
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/tests/rtsan_test_interceptors_posix.cpp:391]: This test (and the one above) exists to ensure we pass on the arg correctly. It's just a simple "lights on" test of ioctl getting info from a socket

File: compiler-rt/lib/rtsan/rtsan_interceptors_posix.cpp [hunk 2]
```diff
@@ -190,6 +190,23 @@ INTERCEPTOR(int, fcntl, int filedes, int cmd, ...) {
...
+  using arg_type = intptr_t;
+  static_assert(sizeof(arg_type) >= sizeof(struct ifreq *));
+  static_assert(sizeof(arg_type) >= sizeof(int));
+
+  va_list args;
+  va_start(args, request);
+  arg_type arg = va_arg(args, arg_type);
+  va_end(args);
+
+  return REAL(ioctl)(filedes, request, arg);
```
Comment (vitalybuka) [inline compiler-rt/lib/rtsan/rtsan_interceptors_posix.cpp:207]: is there va_list overload?

Comment (vitalybuka) [reply compiler-rt/lib/rtsan/rtsan_interceptors_posix.cpp:207]: we do these ugly macro, but we know the size of va_list
https://github.com/llvm/llvm-project/blob/8ffe63fb556915c041e8e9bc2d1bf4325f12ba26/compiler-rt/lib/asan/asan_interceptors.cpp#L375-L381

**Review by vitalybuka — APPROVED** [review]

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors_posix.cpp [hunk 3]
```diff
@@ -373,6 +375,57 @@ class RtsanOpenedFileTest : public RtsanFileTest {
...
+  // These initial checks just see if we CAN run these tests.
+  // If we can't (can't open a socket, or can't find an interface, just
+  // gracefully skip.
+  int sock = socket(AF_INET, SOCK_STREAM, 0);
+  if (sock == -1) {
+    perror("socket");
```
Comment (davidtrevelyan) [inline compiler-rt/lib/rtsan/tests/rtsan_test_interceptors_posix.cpp:397]: `perror` is new to me, so I may have misunderstood - but are they really needed? If you're not aware already, you can stream to `GTEST_SKIP` like `GTEST_SKIP() << "socket failed to open"`

Also, no code should be executed after `GTEST_SKIP()`, so the `return;`s can, I believe, be stripped out for more clarity

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors_posix.cpp [hunk 4]
```diff
@@ -373,6 +375,57 @@ class RtsanOpenedFileTest : public RtsanFileTest {
...
+
+  struct ifreq ifr {};
+  strncpy(ifr.ifr_name, ifaddr->ifa_name, IFNAMSIZ - 1);
+
+  int retval = ioctl(sock, SIOCGIFADDR, &ifr);
+  if (retval == -1) {
+    perror("ioctl");
+    close(sock);
+    freeifaddrs(ifaddr);
+    ASSERT_TRUE(false) << "ioctl failed";
```
Comment (davidtrevelyan) [inline compiler-rt/lib/rtsan/tests/rtsan_test_interceptors_posix.cpp:418]: Could replace `ASSERT_TRUE(false)` with `FAIL()`, and `return` underneath is also not needed

**Review by davidtrevelyan — APPROVED** [review]

**Pull request merged**